rebalance/rebalance.py

```diff
@@ -168,16 +174,35 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi = Non
     description = "%s to %s" % (outgoing_scid, incoming_scid)
     invoice = plugin.rpc.invoice(msatoshi, label, description, retry_for + 60)
     payment_hash = invoice['payment_hash']
-    success_msg = ""
-    try:
-        excludes = []
-        # excude all own channels to prevent unwanted shortcuts [out,mid,in]
-        mychannels = plugin.rpc.listchannels(source=my_node_id)['channels']
-        for channel in mychannels:
-            excludes += [channel['short_channel_id'] + '/0', channel['short_channel_id'] + '/1']
 
+    rpc_result = None
+    count = 0
+    excludes = [my_node_id]  # excude all own channels to prevent shortcuts
+    nodes = {}               # here we store erring node counts
+    maxhops = 1              # retry higher when getroute fails
+    msat_factor = 4
+
+    try:
         while int(time.time()) - start_ts < retry_for and not plugin.rebalance_stop:
-            r = plugin.rpc.getroute(incoming_node_id, msatoshi, riskfactor=1, cltv=9, fromid=outgoing_node_id, exclude=excludes)
+            count += 1
+            try:
+                r = plugin.rpc.getroute(incoming_node_id,
+                                        fromid=outgoing_node_id,
+                                        exclude=excludes,
+                                        msatoshi=msatoshi * min(maxhops, msat_factor),
+                                        maxhops=maxhops,
+                                        riskfactor=10, cltv=9)
+            except RpcError as e:
+                # could not find route -> change params and restart loop
+                if e.method == "getroute" and e.error.get('code') == 205:
+                    maxhops = maxhops + 1
+                    if maxhops == 5:    # set msatoshi factor to neutral
+                        msat_factor = 1
+                        continue
+                    if maxhops == 10:
+                        rpc_result = {'status':'error', 'message': 'No suitable routes found'}
+                        return cleanup(plugin, label, payload, rpc_result)
```

Comment:
Question about this: Do your numbers include the `route_in` and `route_out` first and last hop we add manually? Because we do `getroute` only for the path in between

Reply:
Yes, my numbers originate from this message:
```
success_msg = {"sent": msatoshi + fees, "received": msatoshi, "fee": fees, "hops": len(route),
               "outgoing_scid": outgoing_scid, "incoming_scid": incoming_scid, "status": "settled",
               "message": f"{msatoshi + fees} sent over {len(route)} hops to rebalance {msatoshi}"}
```
The `route_in` and `route_out` is included.

Reply:
@gallizoltan also, how do you think about `riskfactor`? I set it to `10` where it was `1` iirc, as 10 is also used for normal sendpay.

I tried around with it and recognized that if we use very low values (0 or 1) it tends to return longer routes with bigger `delay` values, which is to be expected, there seems to be some inconsistencies (maybe by design):

Low risk values tend to suppress certain possible short routes, see for example (retry on your own with some ID, you will figure out soon):
```
# try to find a route with low riskfactor and small maxhops
lightning-cli getroute id=02843833374b6c42daa4b780df2066a9985cb8cf97bc4566dd63cf693d32cbf7b5 msatoshi=0.001btc riskfactor=0 maxhops=3
{
   "code": 205,
   "message": "Could not find a route"
}

# same with higher riskfactor gives a route:
lightning-cli getroute id=02843833374b6c42daa4b780df2066a9985cb8cf97bc4566dd63cf693d32cbf7b5 msatoshi=0.001btc riskfactor=10 maxhops=3
{
   "route": [
      {
         "id": "0298f6074a454a1f5345cb2a7c6f9fce206cd0bf675d177cdbf0ca7508dd28852f",
         "channel": "685285x349x0",
         "direction": 1,
         "msatoshi": 100003700,
         "amount_msat": "100003700msat",
         "delay": 63,
         "style": "tlv"
      },
      {
         "id": "03c2abfa93eacec04721c019644584424aab2ba4dff3ac9bdab4e9c97007491dda",
         "channel": "579764x2740x0",
         "direction": 0,
         "msatoshi": 100003500,
         "amount_msat": "100003500msat",
         "delay": 29,
         "style": "tlv"
      },
      {
         "id": "02843833374b6c42daa4b780df2066a9985cb8cf97bc4566dd63cf693d32cbf7b5",
         "channel": "564229x2034x0",
         "direction": 1,
         "msatoshi": 100000000,
         "amount_msat": "100000000msat",
         "delay": 9,
         "style": "tlv"
      }
   ]
}
```

Normally I would assume that it is harder to find routes with a higher rather than lower `riskfactor` value...

Reply:
Ok, I tell you the story of `riskfactor`.
I was writing these parts of the code 2 years ago, and honestly, I had no idea what is `riskfactor`. So I've read the manual, which says ['1 is a conservative value for a stable lightning network with very few failures.'](https://github.com/ElementsProject/lightning/blob/master/doc/lightning-getroute.7.md#recommended-riskfactor-values)
This sounded comforting to me. So I've decided to `riskfactor=1` can be a good choice. I've tried it, and it actually worked. And I was happy about it. 

I didn't try any other values beyond 1. Feel free to optimize it!